instrumentation/java-concurrent/src/main/java/io/opentelemetry/instrumentation/auto/javaconcurrent/KotlinProbeInstrumentation.java

```diff
@@ -0,0 +1,196 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.instrumentation.auto.javaconcurrent;
+
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static net.bytebuddy.matcher.ElementMatchers.takesArguments;
+
+import com.google.auto.service.AutoService;
+import io.grpc.Context;
+import io.opentelemetry.javaagent.tooling.Instrumenter;
+import java.util.HashMap;
+import java.util.Map;
+import kotlin.coroutines.Continuation;
+import kotlin.coroutines.CoroutineContext;
+import kotlin.jvm.functions.Function2;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+@AutoService(Instrumenter.class)
+public class KotlinProbeInstrumentation extends Instrumenter.Default {
+  /*
+  Kotlin coroutines with suspend functions are a form of cooperative "userland" threading
+  (you might also know this pattern as "fibers" or "green threading", where the OS/kernel-level thread
+  has no idea of switching between tasks.  Fortunately kotlin exposes hooks for the key events: knowing when
+  coroutines are being created, when they are suspended (swapped out/inactive), and when they are resumed (about to
+  run again).
+
+  Without this instrumentation, heavy concurrency and usage of kotlin suspend functions will break causality
+  and cause nonsensical span parents/context propagation.
+
+  The basic strategy here is:
+  1) Use the DebugProbes callbacks to learn about coroutine create, resume, and suspend operations
+  2) Wrap the creation Coroutine and its Context and use that wrapping to add an extra Context "key"
+  3) Use the callback for resume and suspend to manipulate our context "key" whereby an appropriate state
+     object can be found (tied to the chain of Continutations in the Coroutine).
+  4) Do our swapping-context dance with that appropriate state
+  5) Test with highly concurrent well-known span causality and ensure everything looks right.
+     Without this instrumentation, this test fails with concurrency=2; with this instrumentation,
+     it passes with concurrency=200.
+
+  I've run the basic shape of this by an expert in kotlin coroutines (Roman Elizarov) and he agreed that the
+  overhead should be fine for production usage.
+  */
+
+  public KotlinProbeInstrumentation() {
+    super("kotlin-coroutines");
+  }
+
+  @Override
+  public ElementMatcher<? super TypeDescription> typeMatcher() {
+    return named("kotlin.coroutines.jvm.internal.DebugProbesKt");
```

Comment:
Just to confirm, I guess there's no public entrypoint that is similar to this one? For example, we won't have any breakage across kotlin version if we can stick to public APIs than internal ones. One API that comes to mind is `startCoroutine`

https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.coroutines/start-coroutine.html

Perhaps my other suggestions can bring to light an idea that can keep interception to the public API. But if not, it's ok too we have other instrumentation reliant on private APIs and use muzzle to bite us when something breaks.

Reply:
You raise some good points.  Yes, we're helped by muzzle here.  We're also helped in that kotlin has a debugging extension using this code (hence the name `DebugProbes`) for tracking coroutine causality, and that their implementation also uses bytecode instrumentation to swap in the debugging version.  Because of the friction introduced by that design, it's unlikely that they'll change this interface frequently 😄